File: csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private int speed;
    private int drainRate;
    private int travel = 0;
    private int battery = 100;
    
    // TODO: define the constructor for the 'RemoteControlCar' class
    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.drainRate = batteryDrain;
    }

    public bool BatteryDrained() => this.battery < this.drainRate;


    public int DistanceDriven() => this.travel;

    public void Drive()
    {
        if (this.battery >= this.drainRate) {
            this.travel += this.speed;
            this.battery -= this.drainRate;
        }
    }

    public static RemoteControlCar Nitro() => new RemoteControlCar(50, 4);

}

class RaceTrack
{
    private int distance;
    
    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        while (!car.BatteryDrained() && car.DistanceDriven() < this.distance) {
            car.Drive();
        }

        return car.DistanceDriven() >= this.distance;
    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int speed;
    private int batteryDrain;
    private int battery = 100;
    private int meters;

    NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return battery < batteryDrain;
    }

    public int distanceDriven() {
        return meters;
    }

    public void drive() {
        if (this.batteryDrained()) {
            return;
        }
        meters += speed;
        battery -= batteryDrain;
    }

    public boolean canDrive(int range) {
        return range <= speed * (100 / batteryDrain);
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;
    
    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        return car.canDrive(distance);
    }
}
